cmd/frontend/graphqlbackend/graphqlutil/connection_resolver.go

```diff
@@ -0,0 +1,233 @@
+
+	if r.args.First != nil {
+		limit := int32(applyMaxPageSize(*r.args.First)) + 1
+		paginationArgs.First = &limit
+	} else if r.args.Last != nil {
+		limit := int32(applyMaxPageSize(*r.args.Last)) + 1
+		paginationArgs.Last = &limit
+	} else {
+		limit := MAX_PAGE_SIZE + 1
+		paginationArgs.First = &limit
+	}
+
+	if r.args.After != nil {
+		after, err := r.store.UnMarshalCursor(*r.args.After)
+		if err != nil {
+			return nil, err
+		}
+
+		paginationArgs.After = after
+	}
+
+	if r.args.Before != nil {
+		before, err := r.store.UnMarshalCursor(*r.args.Before)
+		if err != nil {
+			return nil, err
+		}
+
+		paginationArgs.Before = before
+	}
+
+	return &paginationArgs, nil
+}
+
+func (r *ConnectionResolver[N]) TotalCount(ctx context.Context) (int32, error) {
+	r.once.total.Do(func() {
+		r.data.total, r.data.totalError = r.store.ComputeTotal(ctx)
+	})
+
+	if r.data.totalError != nil || r.data.total == nil {
+		return 0, r.data.totalError
+	}
+
+	return *r.data.total, r.data.totalError
+}
+
+func (r *ConnectionResolver[N]) Nodes(ctx context.Context) ([]*N, error) {
+	r.once.nodes.Do(func() {
+		paginationArgs, err := r.paginationArgs()
+		if err != nil {
+			r.data.nodesError = err
+			return
+		}
+
+		r.data.nodes, r.data.nodesError = r.store.ComputeNodes(ctx, paginationArgs)
+
+		/* NOTE(naman): with `last` argument the items are sorted in opposite
+		 * direction in the SQL query. Here we are reversing the list to return
+		 * them in correct order, to reduce complexity */
+		if r.args.Last != nil {
+			for i, j := 0, len(r.data.nodes)-1; i < j; i, j = i+1, j-1 {
+				r.data.nodes[i], r.data.nodes[j] = r.data.nodes[j], r.data.nodes[i]
+			}
+		}
+	})
+
+	nodes := r.data.nodes
+
+	/* NOTE(naman): we pass actual_limit + 1 to SQL query so that we
+	 * can check for `hasNextPage`. Here we need to remove the extra item,
+	 * last item in case of `first` and first item in case of `last` as
+	 * they are sorted in opposite directions in SQL query.*/
+	if len(nodes) > r.args.Limit() {
+		if r.args.Last != nil {
+			nodes = nodes[1:]
+		} else {
+			nodes = nodes[:len(nodes)-1]
+		}
+	}
+
+	return nodes, r.data.nodesError
+}
+
+func (r *ConnectionResolver[N]) PageInfo(ctx context.Context) (*ConnectionPageInfo[N], error) {
+	nodes, err := r.Nodes(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ConnectionPageInfo[N]{
+		len(r.data.nodes),
+		nodes,
+		r.store,
+		r.args,
+	}, nil
+}
+
+type ConnectionPageInfo[N ConnectionNode] struct {
+	fetchedNodesCount int
+	nodes             []*N
+	store             ConnectionResolverStore[N]
+	args              *ConnectionResolverArgs
+}
+
+func (p *ConnectionPageInfo[N]) HasNextPage() bool {
+	if p.args.Before != nil {
+		return true
+	}
+
+	return p.fetchedNodesCount > p.args.Limit()
+}
+
+func (p *ConnectionPageInfo[N]) HasPreviousPage() bool {
+	if p.args.After != nil {
+		return true
+	}
+
+	if p.args.Before != nil {
+		return p.fetchedNodesCount > p.args.Limit()
+	}
+
+	return false
+}
```

Comment:
This logic works for all cases I think but it can be similfied a bit to the inverse of my above:

```suggestion
func (p *ConnectionPageInfo[N]) HasNextPage() bool {
	if p.args.First != nil {
		return p.fetchedNodesCount > p.args.Limit()
	}
	return p.args.Before != nil
}
```

Reply:
resolved.